src/cpp/preprocessor/preprocessor_input.h

```diff
@@ -59,6 +61,16 @@ class preprocessor_input
   {
     m_sym.insert( m_sym.end(), syms.begin(), syms.end());
   }
+
+  void add_metadata(const std::string& key, const std::string& val)
+  {
+    m_metadata[key] = val;
+  }
+
+  std::unordered_map<std::string, std::string>& get_metadata()
```

Comment:
This is my personal taste, but when scanning the code it is very hard to locate the function name, when the return type is a complex as here.  I prefer keeping the important things at the left margin:
```suggestion
  std::unordered_map<std::string, std::string>& 
  get_metadata()
```